File: 10-code-control/12-flag-variations/node/12-flag-variations.js
#!/usr/bin/env node
/** Serve the flag-variations grid navigator web UI on a local HTTP server. */

const http = require("http");
const fs = require("fs");
const path = require("path");
const LaunchDarkly = require("@launchdarkly/node-server-sdk");
const { evaluateFlags } = require("../flag-variations");
const { detectHostOs, HOST_OS_ATTR } = require("../host-os");
const {
  api_config,
  listFlagControls,
  applyFlagControl,
} = require("../ld-flag-controls");

// LaunchDarkly capability: Multivariate flag evaluation + anonymous contexts
// See: https://launchdarkly.com/docs/sdk/features/flag-types
// See: https://launchdarkly.com/docs/sdk/features/anonymous

const PORT = Number(process.env.PORT || 8080);
const ROOT = __dirname;
const HOST_OS = detectHostOs();

let ldClient = null;

async function initLaunchDarkly() {
  const sdkKey = process.env.LD_SDK_KEY;
  if (!sdkKey) {
    console.warn("Warning: LD_SDK_KEY not set — flags use defaults.");
    return;
  }
  ldClient = LaunchDarkly.init(sdkKey, { privateAttributes: [HOST_OS_ATTR] });
  try {
    await ldClient.waitForInitialization({ timeout: 5 });
  } catch (err) {
    console.warn("Warning: LaunchDarkly SDK did not initialize — flags use defaults.");
    ldClient.close();
    ldClient = null;
  }
}

function sendJson(res, status, body) {
  const payload = JSON.stringify(body);
  res.writeHead(status, {
    "Content-Type": "application/json; charset=utf-8",
    "Content-Length": Buffer.byteLength(payload),
    "Cache-Control": "no-store",
  });
  res.end(payload);
}

function readJson(req) {
  return new Promise((resolve, reject) => {
    let raw = "";
    req.setEncoding("utf8");
    req.on("data", (chunk) => {
      raw += chunk;
      if (raw.length > 1024 * 1024) {
        reject(new Error("Request body is too large"));
        req.destroy();
      }
    });
    req.on("end", () => {
      try {
        const parsed = JSON.parse(raw || "{}");
        if (!parsed || Array.isArray(parsed) || typeof parsed !== "object") {
          throw new Error("Request body must be a JSON object");
        }
        resolve(parsed);
      } catch (error) {
        reject(
          error instanceof SyntaxError
            ? new Error("Request body must be JSON")
            : error
        );
      }
    });
    req.on("error", reject);
  });
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, `http://127.0.0.1:${PORT}`);

  if (req.method === "GET" && url.pathname === "/api/flags") {
    const username = (url.searchParams.get("username") || "").trim();
    if (!username) {
      sendJson(res, 400, { error: "username query parameter is required" });
      return;
    }
    sendJson(res, 200, await evaluateFlags(ldClient, username, HOST_OS));
    return;
  }

  if (req.method === "GET" && url.pathname === "/api/flag-controls") {
    sendJson(res, 200, await listFlagControls());
    return;
  }

  if (req.method === "GET" && url.pathname === "/api/bootstrap") {
    sendJson(res, 200, {
      appBanner: "12-flag-variations[node]",
      hostOs: HOST_OS,
      controls: api_config(),
    });
    return;
  }

  if (req.method === "POST" && url.pathname === "/api/flag-controls") {
    try {
      const body = await readJson(req);
      const key = String(body.key || "").trim();
      if (!key) throw new Error('"key" is required');
      const options = {};
      if (Object.prototype.hasOwnProperty.call(body, "on")) {
        if (typeof body.on !== "boolean") throw new Error('"on" must be a boolean');
        options.on = body.on;
      }
      if (Object.prototype.hasOwnProperty.call(body, "fallthrough")) {
        options.fallthrough = body.fallthrough;
      }
      sendJson(res, 200, await applyFlagControl(key, options));
    } catch (error) {
      const status = /required|must be|not allowed|Provide|No variation/.test(error.message)
        ? 400
        : 502;
      sendJson(res, status, { ok: false, error: error.message || String(error) });
    }
    return;
  }

  const urlPath = url.pathname === "/" ? "/index.html" : url.pathname;
  const filePath = path.join(ROOT, urlPath);

  if (!filePath.startsWith(ROOT)) {
    res.writeHead(403);
    res.end("Forbidden");
    return;
  }

  fs.readFile(filePath, (err, data) => {
    if (err) {
      res.writeHead(404);
      res.end("Not found");
      return;
    }
    const ext = path.extname(filePath);
    const type = ext === ".html" ? "text/html" : "text/plain";
    res.writeHead(200, { "Content-Type": type });
    res.end(data);
  });
});

initLaunchDarkly().then(() => {
  server.listen(PORT, "127.0.0.1", () => {
    console.log(`Grid navigator (flag variations) running at http://127.0.0.1:${PORT}/`);
    console.log("Press Ctrl+C to stop.");
  });
});

process.on("SIGINT", () => {
  if (ldClient) ldClient.close();
  process.exit(0);
});

File: 10-code-control/12-flag-variations/python/12-flag-variations.py
#!/usr/bin/env python3
"""Serve the flag-variations grid navigator web UI on a local HTTP server."""

from __future__ import annotations

import json
import os
import sys
from http.server import SimpleHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import parse_qs, urlparse

import ldclient
from ldclient import Config
from ldclient.client import LDClient

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))
from flag_variations import evaluate_flags as eval_flags  # noqa: E402
from host_os import (  # noqa: E402
    ANONYMOUS_CONTEXT_KEY,
    FLAG_ANON_OS_EMOJI,
    HOST_OS_ATTR,
    detect_host_os,
)
from ld_flag_controls import (  # noqa: E402
    api_config,
    apply_flag_control,
    list_flag_controls,
)

# LaunchDarkly: multivariate evaluation (string / number / JSON) + anonymous boolean
# https://launchdarkly.com/docs/sdk/features/flag-types
# In-app Controls use the REST API (not the SDK).
# https://launchdarkly.com/docs/api/feature-flags/patch-feature-flag

ROOT = Path(__file__).parent
_ld_client: LDClient | None = None
_host_os = detect_host_os()


def init_launchdarkly() -> None:
    global _ld_client
    sdk_key = os.environ.get("LD_SDK_KEY")
    if not sdk_key:
        print("Warning: LD_SDK_KEY not set — flags use defaults.", flush=True)
        return
    ldclient.set_config(Config(sdk_key, private_attributes=[HOST_OS_ATTR]))
    _ld_client = ldclient.get()
    if not _ld_client.is_initialized():
        print("Warning: LaunchDarkly SDK did not initialize — flags use defaults.", flush=True)


def evaluate_flags(username: str) -> dict:
    """Evaluate flags and attach LD context info for the lab Context tab."""
    response = eval_flags(_ld_client, username, _host_os)
    response["ldContext"] = {
        "user": {
            "kind": "user",
            "key": username,
            "note": "String, number, and JSON flags evaluate against this user context.",
        },
        "anonymous": {
            "kind": "user",
            "key": ANONYMOUS_CONTEXT_KEY,
            "anonymous": True,
            "attributes": {HOST_OS_ATTR: _host_os},
            "privateAttributes": [HOST_OS_ATTR],
            "flagKey": FLAG_ANON_OS_EMOJI,
            "note": (
                f"{FLAG_ANON_OS_EMOJI} uses this anonymous context. "
                f"{HOST_OS_ATTR} is private (targeting only; redacted from analytics)."
            ),
        },
    }
    return response


def _json_response(handler: SimpleHTTPRequestHandler, status: int, payload: object) -> None:
    body = json.dumps(payload).encode("utf-8")
    handler.send_response(status)
    handler.send_header("Content-Type", "application/json; charset=utf-8")
    handler.send_header("Content-Length", str(len(body)))
    handler.send_header("Cache-Control", "no-store")
    handler.end_headers()
    handler.wfile.write(body)


def _read_json(handler: SimpleHTTPRequestHandler) -> dict:
    length = int(handler.headers.get("Content-Length") or "0")
    raw = handler.rfile.read(length) if length else b"{}"
    try:
        data = json.loads(raw.decode("utf-8") or "{}")
    except json.JSONDecodeError as exc:
        raise ValueError("Request body must be JSON") from exc
    if not isinstance(data, dict):
        raise ValueError("Request body must be a JSON object")
    return data


class Handler(SimpleHTTPRequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=str(ROOT), **kwargs)

    def do_GET(self) -> None:
        parsed = urlparse(self.path)
        if parsed.path == "/api/flags":
            params = parse_qs(parsed.query)
            username = (params.get("username") or [""])[0].strip()
            if not username:
                self.send_error(400, "username query parameter is required")
                return
            _json_response(self, 200, evaluate_flags(username))
            return

        if parsed.path == "/api/flag-controls":
            _json_response(self, 200, list_flag_controls())
            return

        if parsed.path == "/api/bootstrap":
            _json_response(
                self,
                200,
                {
                    "appBanner": "12-flag-variations[python]",
                    "hostOs": _host_os,
                    "controls": api_config(),
                },
            )
            return

        super().do_GET()

    def do_POST(self) -> None:
        parsed = urlparse(self.path)
        if parsed.path != "/api/flag-controls":
            self.send_error(404, "Not found")
            return
        try:
            data = _read_json(self)
            key = str(data.get("key") or "").strip()
            if not key:
                raise ValueError('"key" is required')
            turn_on: bool | None = bool(data["on"]) if "on" in data else None
            fallthrough = data["fallthrough"] if "fallthrough" in data else None
            result = apply_flag_control(key, turn_on=turn_on, fallthrough=fallthrough)
            _json_response(self, 200, result)
        except ValueError as exc:
            _json_response(self, 400, {"ok": False, "error": str(exc)})
        except Exception as exc:  # noqa: BLE001
            _json_response(self, 502, {"ok": False, "error": str(exc)})


def main() -> None:
    init_launchdarkly()
    cfg = api_config()
    if cfg["configured"]:
        print(
            f"Flag controls ready (project={cfg['projectKey']} env={cfg['environmentKey']})",
            flush=True,
        )
    else:
        print(
            "Warning: flag Controls disabled — missing " + ", ".join(cfg["missing"]),
            flush=True,
        )

    # PORT override: solo default 8080; series portal typically uses 8120.
    port = int(os.environ.get("PORT") or "8080")
    server = ThreadingHTTPServer(("127.0.0.1", port), Handler)
    print(f"Grid navigator (flag variations) running at http://127.0.0.1:{port}/")
    print("Press Ctrl+C to stop.")
    try:
        server.serve_forever()
    finally:
        if _ld_client is not None:
            _ld_client.close()


if __name__ == "__main__":
    main()
